modules/apps/bean-portlet/bean-portlet-registration-impl/src/main/java/com/liferay/bean/portlet/registration/internal/BeanPortletRegistrarImpl.java

```diff
@@ -334,6 +340,83 @@ else if (ArrayUtil.contains(
 		}
 	}
 
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTracker = new ServiceTracker<>(
+			bundleContext, BeanPortletRegistrarBag.class,
+			new ServiceTrackerCustomizer
+				<BeanPortletRegistrarBag, BeanPortletRegistrarBag>() {
+
+				@Override
+				public BeanPortletRegistrarBag addingService(
+					ServiceReference<BeanPortletRegistrarBag>
+						serviceReference) {
+
+					BeanPortletRegistrarBag beanPortletRegistrarBag =
+						bundleContext.getService(serviceReference);
+
+					_serviceRegistrations.put(
+						beanPortletRegistrarBag,
+						register(
+							beanPortletRegistrarBag.
+								getBeanFilterMethodFactory(),
+							beanPortletRegistrarBag.
+								getBeanFilterMethodInvoker(),
+							beanPortletRegistrarBag.
+								getBeanPortletMethodFactory(),
+							beanPortletRegistrarBag.
+								getBeanPortletMethodInvoker(),
+							beanPortletRegistrarBag.getDiscoveredClasses(),
+							beanPortletRegistrarBag.getServletContext()));
+
+					return beanPortletRegistrarBag;
+				}
+
+				@Override
+				public void modifiedService(
+					ServiceReference<BeanPortletRegistrarBag> serviceReference,
+					BeanPortletRegistrarBag beanPortletRegistrarBag) {
+				}
+
+				@Override
+				public void removedService(
+					ServiceReference<BeanPortletRegistrarBag> serviceReference,
+					BeanPortletRegistrarBag beanPortletRegistrarBag) {
+
+					bundleContext.ungetService(serviceReference);
+
+					unregister(
+						_serviceRegistrations.remove(beanPortletRegistrarBag),
+						beanPortletRegistrarBag.getServletContext());
+				}
+
+			});
+
+		_serviceTracker.open();
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_serviceTracker.close();
```

Comment:
@W2jx `ServiceTracker.close()` calls `removedService` for all the tracked items. Please double check whether the for-loop is still needed.

Reply:
Hi @dantewang , It's my bad. I reviewed the logic of method `ServiceTracker.close()` again, it will call method `ServiceTracker.customizerRemoved`, and method `customizer.removedService` will be called in it. Therefore, there is no need to clean up the map separately. I will resend it.